### PR TITLE
sys/btree: Initial include of a self-balancing binary search tree module

### PR DESCRIPTION
**Pull request by bergzand**
### Contribution description

This PR adds a simple binary search tree module to RIOT. It is optimized for low flash footprint and fast key lookups. Without the auxiliary debug functions the flash footprint is approx 600 bytes on a samr21-xpro.

### Testing procedure

A test is supplied. Please also read the accompanied docs and check whether the module usage is clear.

### Issues/PRs references

None

### CODE REVIEW

File: sys/btree/btree.c [hunk 1]
```diff
@@ -0,0 +1,320 @@
...
+    _find_key(node, &min, 0);
+    return min;
+}
+
+static btree_node_t *_find_max(btree_node_t *node)
+{
+    btree_node_t *max = NULL;
+    _find_key(node, &max, UINT32_MAX);
+    return max;
+}
```
Comment (benpicco) [inline sys/btree/btree.c:89]: Won't those return `NULL` most of the time?
`_find_key()` only returns exact matches.

File: sys/btree/btree.c [hunk 2]
```diff
@@ -0,0 +1,320 @@
...
+ * Copyright (C) 2020 Inria
+ * Copyright (C) 2020 Koen Zandberg <koen@bergzand.net>
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup sys_btree2
```
Comment (benpicco) [inline sys/btree/btree.c:11]: That group is not defined.

File: sys/include/btree.h [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+ * @brief Auxiliary function to dump a print of the tree to stdio
+ *
+ * @param   btree   tree to print
+ */
+void btree_dump(btree_t *btree);
+
+#ifdef __cplusplus
+}
+#endif
+#endif /* UUID_H */
```
Comment (benpicco) [inline sys/include/btree.h:204]:
```suggestion
#endif /* BTREE_H */
```

File: sys/btree/btree.c [hunk 4]
```diff
@@ -0,0 +1,320 @@
...
+{
+    int child_balance = 0;
+    while (node) {
+        /* Determine max length of the other branch */
+        /* TODO: reuse the result from the previous loop */
+        int balance = _max_path(node->right) - _max_path(node->left);
+        if (balance > 1) {
+            /* right/left case */
+            if (child_balance < 0) {
+                /* Tranform to right/right case */
```
Comment (benpicco) [inline sys/btree/btree.c:146]:
```suggestion
                /* Transform to right/right case */
```

File: sys/btree/btree.c [hunk 5]
```diff
@@ -0,0 +1,320 @@
...
+            }
+            /* right/right case */
+            _rotate_left(btree, node, node->right);
+            node = node->right;
+            child_balance = 0;
+        }
+        else if (balance < -1) {
+            /* left/right case */
+            if (child_balance > 0) {
+                /* Tranform to left/left case */
```
Comment (benpicco) [inline sys/btree/btree.c:158]:
```suggestion
                /* Transform to left/left case */
```

**Comment by fjmolinas** [comment]
ping @bergzand !

**Comment by fjmolinas** [comment]
@bergzand are you still interested by this one?

**Pull request closed without merging**